src/PrintCore/PrintCore.cs

```diff
@@ -0,0 +1,775 @@
+
+		// Return is overloaded - if negative, a PMStatusCode.
+		public PMPrinterState PrinterState {
+			get {
+				PMPrinterState s;
+				var code = PMPrinterGetState (handle, out s);
+				if (code == PMStatusCode.Ok)
+					return s;
+
+				return (PMPrinterState) code;
+			}
+		}
+		
+		[DllImport (Constants.PrintCoreLibrary)]
+		extern static PMStatusCode PMPrinterGetMimeTypes (IntPtr printer, IntPtr settings, out IntPtr arrayStr);
+
+		public PMStatusCode TryGetMimeTypes (PMPrintSettings settings, out string [] mimeTypes)
+		{
+			IntPtr m;
+			var code = PMPrinterGetMimeTypes (handle, settings == null ? IntPtr.Zero : settings.Handle, out m);
+			if (code != PMStatusCode.Ok){
+				mimeTypes = null;
+				return code;
+			}
+			mimeTypes = PMPrintSession.FetchArray (m, owns: false);
+			return PMStatusCode.Ok;
+		}
+
+		[DllImport (Constants.PrintCoreLibrary)]
+		extern static PMStatusCode PMPrinterGetPaperList (IntPtr printer, out IntPtr arrayStr);
+		public PMStatusCode TryGetPaperList (out PMPaper [] paperList)
+		{
+			IntPtr m;
+			var code = PMPrinterGetPaperList (handle, out m);
+			if (code != PMStatusCode.Ok){
+				paperList = null;
+				return code;
+			}
+			int c = (int) CFArray.GetCount (m);
+			paperList = new PMPaper [c];
+			for (int i = 0; i < c; i++)
+				paperList [i] = new PMPaper (CFArray.CFArrayGetValueAtIndex (m, i), owns: false);
+
+			return PMStatusCode.Ok;
+		}
+
+		public PMPaper [] PaperList {
+			get {
+				IntPtr m;
+				if (PMPrinterGetPaperList (handle, out m) != PMStatusCode.Ok)
+					return new PMPaper [0];
+
+				int c = (int) CFArray.GetCount (m);
+				var paperList = new PMPaper [c];
+				for (int i = 0; i < c; i++)
+					paperList [i] = new PMPaper (CFArray.CFArrayGetValueAtIndex (m, i), owns: false);
+
+				return paperList;
+			}
+		}
+
+		[DllImport (Constants.PrintCoreLibrary)]
+		extern static PMStatusCode PMPrinterPrintWithFile (IntPtr handle, IntPtr settings, IntPtr pageFormat, IntPtr strMimeType, IntPtr fileUrl);
+
+		public PMStatusCode TryPrintFile (PMPrintSettings settings, PMPageFormat pageFormat, NSUrl fileUrl, string mimeType = null)
+		{
+			if (settings == null)
+				throw new ArgumentNullException (nameof (settings));
+			if (fileUrl == null)
+				throw new ArgumentNullException (nameof (fileUrl));
+				    
+			IntPtr mime = mimeType == null ? IntPtr.Zero : CFString.LowLevelCreate (mimeType);
+			var code = PMPrinterPrintWithFile (handle, settings.handle, pageFormat == null ? IntPtr.Zero : pageFormat.handle, mime, fileUrl.Handle);
+			if (mime != IntPtr.Zero)
+				CFObject.CFRelease (mime);
+			return code;
+		}
+
+		[DllImport (Constants.PrintCoreLibrary)]
+		extern static PMStatusCode PMPrinterPrintWithProvider (IntPtr printer, IntPtr settings, IntPtr pageFormat, IntPtr strMimeType, IntPtr cgDataProvider);
+
+		public PMStatusCode TryPrintFromProvider (PMPrintSettings settings, PMPageFormat pageFormat, CGDataProvider provider, string mimeType = null)
```

Comment:
`FromProvider` does not seem needed, it cane inferred from the method's parameters